### ISSUE TITLE
[Parquet] Implement Variant type support in Parquet

### ISSUE THREAD

**Issue by alamb** [apache/arrow-rs#6736]
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
Parquet recently adopted the Variant type from Spark: https://github.com/apache/parquet-format/blob/master/VariantEncoding.md

Details on 
- https://github.com/apache/parquet-format/pull/456

**Describe the solution you'd like**
I would like to implement variant support in parquet-rs 

**Describe alternatives you've considered**

**Additional context**
I am not sure if any other parquet implementations have implemented this yet / if there are example parquet files. I will attempt to find out

**Comment by CurtHagenlocher** [comment]
There's an implementation in Spark (try [here](https://github.com/apache/spark/blob/007c31df6da7b741f3a2a43c859ebed9f801dcfa/common/variant/src/main/java/org/apache/spark/types/variant/Variant.java#L43) for starters) but when I last looked ~two months ago there was no binary test data; only some round trips via JSON.

**Comment by laurentgo** [comment]
What would be required on the Arrow side of things? Just an extension, or would we also need methods to access/manipulate the content as well?

**Comment by tustvold** [comment]
I do wonder if a precursor to supporting this would be some way to translate / represent the variant data in arrow, whilst there are non-arrow APIs, they'd likely struggle to accommodate this addition.

**Comment by findepi** [comment]
From arrow perspective, would that be a new DataType, or rather a convention of using DataType::Struct with two Binary fields?

A fully performance variant implementation should be able to leverage file-level column disaggregation (shredding), but I do think this could come as a follow-up to a "normal" Variant type implementation.

**Comment by tustvold** [comment]
> From arrow perspective, would that be a new DataType, or rather a convention of using DataType::Struct with two Binary fields?

I don't know, I've not really been following the variant proposal close enough to weigh in here. However, my understanding is that shredding is one of the major motivators for this getting added to parquet, as without it you might as well just embed any record format, e.g. Avro. I therefore suspect most use-cases will be at least partially shredded, and the reader will need to handle this case. This is especially true given the variant_value is NULL when the data is shredded, as opposed to say duplicating the content (which would have its own issues TBC), and so we can't just ignore the shredded data.

Unfortunately I can't see an obvious way to be able to represent this sort of semi-structured data within the arrow format without introducing a new DataType that is able to accommodate arrays having the same type, but different child layouts...

TLDR I suspect actioning this will require arrow defining a way to represent semi-structured data...

**Comment by findepi** [comment]
There needs to be a way to represent a series of variant values having "no type in common" (variant integer, variant boolean, variant varchar, etc all mixed up). For that some blob-like representation with internal structure seems natural.
Then there should be a way to carry-on the shredded columns without having to put them back into that blob, so yes, one type, different child layouts. 
It feels to me that the runtime representation will end up being similar to what is defined in Parquet (https://github.com/apache/parquet-format/blob/master/VariantShredding.md)... so maybe it should be the same representation to provide for an efficient read path.

**Comment by findepi** [comment]
When considering what to do in Arrow, we should also keep an eye on the ongoing effort in Iceberg https://github.com/apache/iceberg/issues/10392#issuecomment-2512272807
This could inform some design decisions.
cc @Xuanwo

**Comment by alamb** [comment]
There is now a related proposal being added to paruqet (shredding the variant type:

- https://github.com/apache/parquet-format/pull/461

FWIW I would *love* to have some rust representation (aka someone who wanted to implement the variant type in the rust parquet decoder). If you are interested there are relevant conversations going on on the parquet mailing list

**Comment by alamb** [comment]
to be clear, we may not be able to merge / really use variants in rust until it gets into Arrow, but we can work out how it would work in parquet (maybe with the non arrow interfaces) first

**Comment by CurtHagenlocher** [comment]
There's a pure Python implementation in [the Spark repo](https://github.com/apache/spark/blob/master/python/pyspark/sql/variant_utils.py). It's *almost* standalone, having a dependency only on `PySparkValueError`.

**Comment by alamb** [comment]
I spent some time listening and thinking about this on the parquet call yesterday: https://lists.apache.org/thread/cnn6264g56jktrwmplz89x8cgkcvr4ql

Note there is a thread on the arrow mailing list about adding variant support in arrow-rs:
- https://lists.apache.org/thread/lsmkmxsp1qvjzn497z582hjm0w8hmg0n

(and it looks like @wjones127 made some sort of demo using an extension type in datafusion):
* https://github.com/datafusion-contrib/datafusion-functions-variant


> Unfortunately I can't see an obvious way to be able to represent this sort of semi-structured data within the arrow format 

What I suggest is that the parquet reader reads variant columns as `Binary` / `LargeBinary` with an arrow extension type annotation, which would let downstream projects interpret / read the extension type correctly

I think one challenge will be "how to tell the parquet writer to write / annotate the columns as variant"

Before we can do anything useful with the variant type, we'll need a library to parse / interpret a variant value (aka the equivalent of a JSON parser / set of objects)

**Comment by alamb** [comment]
So the first step as I see it is that someone has to code up / find a Rust implementation for working with variant values. This could be a port or inspired by the ones @CurtHagenlocher 's points at here:

> There's a pure Python implementation in [the Spark repo](https://github.com/apache/spark/blob/master/python/pyspark/sql/variant_utils.py). 

> There's an implementation in Spark (try [here](https://github.com/apache/spark/blob/007c31df6da7b741f3a2a43c859ebed9f801dcfa/common/variant/src/main/java/org/apache/spark/types/variant/Variant.java#L43) for starters)

Once we have such a library, we can then figure out if/how it should be used in the parquet reader/writer directly

**Comment by wjones127** [comment]
I haven't had time to work on this recently, but for a rust implementation for working with variant values, anyone should feel free to work off what I had started in the https://github.com/datafusion-contrib/datafusion-functions-variant/tree/main/open-variant repo. There is a core `open-variant` crate there that's just meant to be reading and writing variant values.

**Comment by alamb** [comment]
The current variant spec appears to be
- https://github.com/apache/parquet-format/blob/37b6e8b863fb510314c07649665251f6474b0c11/VariantEncoding.md

**Comment by alamb** [comment]
Since variant is part of the parquet spec now I think the code to interpret could easily belong on the arrow-rs repository and the parquet module. It will be an interesting Rust API design challenge I think to make a really efficient/zero copy encoder/decoder

**Comment by alamb** [comment]
Here is a PR to implement variant in C/C++ from @neilechao
- https://github.com/apache/arrow/pull/45375

**Comment by alamb** [comment]
I have requested example VARIANTs on the mailing list as well:
- https://lists.apache.org/thread/22dvcnm7v5d30slzc3hp8d9qq8syj1dq

**Comment by adriangb** [comment]
Digging deep into some Spark code I found some pretty enlightening information about how this will actually be encoded into Parquet: https://github.com/apache/spark/commit/3c3d1a6a9f6daf6db5148f1423f49f4bce142858#diff-ca9eeead72220965c7bbd52631f7125d4c1ef22b898e5baec83abc7be9495325

So it seems that https://github.com/apache/datafusion/issues/11745 will ultimately be a blocker for proper support.

I think the things we'll need here are:
- Ability to project individual struct fields, in particular `column -> typed_value -> field_name`, for selection and during predicate pushdown pruning
- Functions that operate on the entire structure and know how to parse the binary metadata/value fields
- A type that you can declare at the schema level that doesn't force you to exhaustively define the unknown typed fields of the struct
- Statistics support for nested struct fields

On the DataFusion side I think all we need is something like https://github.com/apache/datafusion/pull/15057 to allow rewriting a filter or projection such as `variant_get(col, 'key') = 5` into `"col.typed_value.key.typed_value" = 5` on a per-file level if we see from the file schema that `a` is shredded. Then if all of the above is in place stats filtering, selecting reading of the column for filtering / projection, etc.

**Comment by alamb** [comment]
Said begging email: https://lists.apache.org/thread/jr0yds4o97rtdkc7dmmsk4ck0odvc9h8

**Comment by alamb** [comment]
I got a response from @cashmand ❤  and I filed an issue in the parquet-testing repo to track the work to add examples
- https://github.com/apache/parquet-testing/issues/75

**Comment by adriangb** [comment]
> Here is an update from [Martin Prammer](https://www.cs.cmu.edu/~mprammer/) (not sure if he has a github handle)

Seems like @mprammer does 😄

**Comment by alamb** [comment]
I just had a discussion with @PinkCrow007 and @PinkCrow007, as I understand it the next steps will be:
1. Create a draft / work in progress PR that we can start reviewing / providing feedback on based on the fork: https://github.com/apache/arrow-rs/compare/main...PinkCrow007:arrow-rs:variant-clean. 

The eventual goal will be to break it up and start merging it in as pieces:
1. Support for reading binary (`&[u8]`) as variants (accessing via fields, etc)
2. Support for reading/writing to parquet
3. SUpport for shredding, etc

It is going to be so great

**Comment by alamb** [comment]
I took a quick skim through the code in https://github.com/apache/arrow-rs/compare/main...PinkCrow007:arrow-rs:variant-clean and I found it easy to understand and well structured. I am very much looking forward to the PR

**Comment by alamb** [comment]
I have been studying the variant spec, and various implementation. It seems variant support to/from json is quite well covered. There are things in the spec (like Time, for example) that are not well represented/implemented in open source spark. I'll. definitely focus on the JSON stuff first
- https://github.com/apache/parquet-testing/issues/75#issuecomment-2800115829

**Comment by alamb** [comment]
I spent a while today writing up some first sugested steps and linked them to the ticket
- [ ] https://github.com/apache/arrow-rs/issues/7423
- [ ] https://github.com/apache/arrow-rs/issues/7424
- [ ] https://github.com/apache/arrow-rs/issues/7425
- [ ] https://github.com/apache/arrow-rs/issues/7426

It would be great if someone wanted to take a crack at 
- https://github.com/apache/arrow-rs/issues/7423

I think that will unblock a lot of the rest of what is going on

I am out the rest of the week but hopefully I can check email occasionally

**Comment by alamb** [comment]
Status report (also reported to the parquet mailing list): https://lists.apache.org/thread/dy22njos6c0wbo82s377wvbobbd7y6lx

I am pretty stoked to report  progress on a Rust Variant implementation (see epic here[0]). 

1. We have added example binary data in parquet-testing[1] (Thanks Micah for the review and merge)
2. Jiaying's prototype [2] is looking good and we are preparing to start merging it in pieces
3. I have created a PR to add a parquet-variant crate to arrow-rs[3] (looking for a review 🎣)

Once we have the parquet-variant crate merged, I expect a series of PRs that incrementally adds support for different parts of the Variant specification. Once this is done, we will move on to integrating it into the parquet decoder. 

Exciting times,
Andrew


[0]: https://github.com/apache/arrow-rs/issues/6736
[1]: https://github.com/apache/parquet-testing/pull/76
[2]: https://github.com/apache/arrow-rs/pull/7452
[3]: https://github.com/apache/arrow-rs/pull/7485

**Comment by findepi** [comment]
> [@findepi](https://github.com/findepi) -- I heard today you may be working on variant support as well. I wonder if you have any thoughts about the above plan (or perhaps already have it implemented and would be willing to share 😆 )

~SDF~ dbt indeed has an internal Variant implementation. I would need to ensure there would be openness to contribute, but I the code wouldn't be directly usable:

- the implementation is aligned with Snowflake VARIANT semantics
- no Parquet-level logic, just in-memory implementation. It's just a binary for storage. In particular no cdisaggregated columns.
- the implementation is pragmatic, but not necessarily well designed -- just a binary blob with Protobof encoded data. Flatbuffers could be more efficient from UDFs perspective

@alamb if this would be nonetheless somehow useful, lemme know and I can do the clearance.

**Comment by alamb** [comment]
> @alamb if this would be nonetheless somehow useful, lemme know and I can do the clearance.

It doesn't sound like it -- I think we are looking for a Variant in this crate that is aligned with the parquet spec

**Comment by mkarbo** [comment]
Hey @alamb and @PinkCrow007, 

We're looking at picking up some of this work on our side, we think it would make sense to take a stab at the suggested path forward by taking #7452, extending it and breaking it into a new PRs targeting the new crate as you suggest @alamb. We're thinking of starting with the Read API.

Happy to coordinate efforts going forward.

**Comment by alamb** [comment]
Thank you @mkarbo

**Comment by alamb** [comment]
> Hi [@mkarbo](https://github.com/mkarbo) and [@alamb](https://github.com/alamb),
> 
> Thanks for the update. Great to see more interest in pushing this forward! Just to clarify: I’m currently working on the create API and encode/decode logic, which also touches parts of the read API to validate the correctness. That said, it makes sense to develop these in parallel — happy to collaborate and coordinate! Also, would it make sense to also assign issues to me?

Sounds great -- thank you so much. I think @mkarbo  is going to try to break out just the read API into a new PR that we'll try and merge fist

Note due to how the github permissions work, I can't assign you to any issues unless you comment on them first. Feel free to comment on them so I can assigned them to you

**Comment by ianmcook** [comment]
There's a discussion happening on the Arrow dev mailing list about adding Variant as a canonical extension type in the Arrow spec. Input from the Arrow Rust and DataFusion developer community would be much appreciated.
- The discussion thread in the mailing list archive: https://lists.apache.org/thread/w06cxdojjcmry4m9vb0bo7owd1jsbtz5
- How to subscribe to the Arrow dev mailing list: https://arrow.apache.org/community/#mailing-lists

**Comment by adriangb** [comment]
I think https://github.com/apache/arrow-rs/issues/5699 will end up being a blocker for this work if we want to support shredded columns. Does anyone have the appetite to pick that up?

**Comment by alamb** [comment]
> I think [#5699](https://github.com/apache/arrow-rs/issues/5699) will end up being a blocker for this work if we want to support shredded columns. Does anyone have the appetite to pick that up?

I added it to the list on this ticket -- thank you.

**Comment by alamb** [comment]
Here is a PR that might be of general interest: Proposing `VariantArray` and `VariantArrayBuilder` APIs
- https://github.com/apache/arrow-rs/pull/7905

**Comment by alamb** [comment]
I spent a few minutes reorganizing the subtasks on the description of this ticket to make it easier to find what we are looking for

**Comment by alamb** [comment]
In case anyone is interested, I have a proposal for how to implement reading of shredded variants here:
- https://github.com/apache/arrow-rs/pull/8021

Any feedback would be most welcome

**Comment by alamb** [comment]
I filed quite a few items under the **Arrow --> Variant conversions** heading above that it would be great to get help with

- [ ] https://github.com/apache/arrow-rs/issues/8049
- [x] https://github.com/apache/arrow-rs/issues/8050
- [ ] https://github.com/apache/arrow-rs/issues/8051
- [ ] https://github.com/apache/arrow-rs/issues/8052
- [ ] https://github.com/apache/arrow-rs/issues/8053
- [ ] https://github.com/apache/arrow-rs/issues/8054
- [ ] https://github.com/apache/arrow-rs/issues/8055
- [ ] https://github.com/apache/arrow-rs/issues/8056
- [x] https://github.com/apache/arrow-rs/issues/8057
- [ ] https://github.com/apache/arrow-rs/issues/8058
- [ ] https://github.com/apache/arrow-rs/issues/8059
- [ ] https://github.com/apache/arrow-rs/issues/8060
- [ ] https://github.com/apache/arrow-rs/issues/8061
- [ ] https://github.com/apache/arrow-rs/issues/8062
- [ ] https://github.com/apache/arrow-rs/issues/8063
- [ ] https://github.com/apache/arrow-rs/issues/8064

**Comment by alamb** [comment]
I filed a bunch more tickets for those of you who might be Variant curious and want to help out

- https://github.com/apache/arrow-rs/issues/8084
- https://github.com/apache/arrow-rs/issues/8082
- https://github.com/apache/arrow-rs/issues/8083
- https://github.com/apache/arrow-rs/issues/8086
- https://github.com/apache/arrow-rs/issues/8087
- https://github.com/apache/arrow-rs/issues/8088

**Comment by alamb** [comment]
I filed a ticket to discuss integrating variant support for real into the `parquet` crate:
- https://github.com/apache/arrow-rs/issues/8132

**Comment by alamb** [comment]
Here is a proposal to add a new feature flag to the parquet crate to support `Variant`: 
- https://github.com/apache/arrow-rs/pull/8133

**Comment by alamb** [comment]
I updated the `shredded_variant` integration test and that PR is now ready for review
- https://github.com/apache/arrow-rs/pull/8325


I also files a bunch of tickets to add support for the additional data types in the tests:
- [ ] https://github.com/apache/arrow-rs/issues/8329
- [ ] https://github.com/apache/arrow-rs/issues/8330
- [ ] https://github.com/apache/arrow-rs/issues/8331
- [ ] https://github.com/apache/arrow-rs/issues/8332
- [ ] https://github.com/apache/arrow-rs/issues/8333
- [ ] https://github.com/apache/arrow-rs/issues/8334
- [ ] https://github.com/apache/arrow-rs/issues/8335
- [ ] https://github.com/apache/arrow-rs/issues/8336
- [ ] https://github.com/apache/arrow-rs/issues/8337

**Comment by alamb** [comment]
Since this ticket is getting out of control with the number of linked issues (it is hard to understand what the current status is and what needs help), what I am thinking we'll do is that once I merge the support for reading/writing the parquet logical type I'll close this ticket as "done" (as we'll have completed the initial implementation) and file a follow on EPIC with remaining work

The major additional areas I think we are missing are support for shredded Lists/Arrays and nested lists/structs, and some additional type support

**Comment by alamb** [comment]
Now that we merged support for reading/writing Variant LogicalTypes to parquet (via the following PR) let'c claim that we have completed the "initial support" for Variant
- https://github.com/apache/arrow-rs/issues/8370

There is still more work to do, so I filed an additional follow on EPIC to try and organize this work
- https://github.com/apache/arrow-rs/issues/8480

A huge shoutout to everyone who helped make this happen (and I am sorry if I missed you)

@PinkCrow007 @mkarbo @scovich @superserious-dev @friendlymatthew @micoo227 @Weijun-H @harshmotw-db @odysa @carpecodeum  @viirya @klion26 @adriangb @kosiew @liamzwbao @codephage2020 @ding-young @mbrobbel @petern48 @sdf-jkl @abacef  and @mprammer

**Comment by HuaHuaY** [comment]
Can we remove WIP state in `parquet-variant/src/lib.rs:27` now?

**Comment by kevinjqliu** [comment]
great catch @HuaHuaY 
https://github.com/apache/arrow-rs/blob/7616e10f12894232eacb807993e55427a067d061/parquet-variant/src/lib.rs#L27-L32
the epic is closed (https://github.com/apache/arrow-rs/issues/6736), so i'd say yes! 

would you like to open a PR to remove those lines?

**Comment by HuaHuaY** [comment]
> would you like to open a PR to remove those lines?

Hi @kevinjqliu, I have opened a PR to remove WIP comments in three files https://github.com/apache/arrow-rs/pull/10238. Please take a look.

**Comment by sdf-jkl** [comment]
@kevinjqliu @HuaHuaY we're still working on finalizing `Variant` support in #8480.

The majority of work is done, but there are still some `breaking` issues that need handling before finalizing `Variant`:
- #8982 and it's sub issues.
- #10145

The rest I'd say are additive and can be done after we make `Variant` production ready.
